### ISSUE TITLE
Incorrect nullability warnings for conditional access in assertion method

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#34875]
**Version Used**: 

Branch master (9 Apr 2019)
Latest commit 665dc2 by Andrew Hall:
Merge pull request #34764 from ryzngard/issue/pullmembers_up_a11y

Pull Members Up UI fixes

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

public class C
{
  public extern void M([NotNull] C? c1, C? c2);

  public void Test1(C? c1, C? c2, C? c3)
  {
    M(c1?.Method(c2 = c1, c1 = null), c3 = c1);

    c1.ToString();
    c2.ToString();
    c3.ToString();
  }

  public extern C? Method(C? c1, C? c2);
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABABgqYEYA6AEQEsIAc1wB7WHwDGULgGERKOAEEahAJ5Q+UANyNGTAMzsATBRmMA3owrtDcAB7wEudgBYKAWQAUAbQByImF8iQgBdUwB+CgkOJAioowBKHQYrG1cKABU4WA5PGUjo2Pz4ooL9BNTLBmtrL2jwrnc4GAALeU8JEwBeKJjeih6CYgTYiUMe6KTdaprorgyRAGUYBD5cQU8pmetO+aWVtY2tmqj9PeXV9c3k6wBfaesDKgdEZ2Km1vbiwrjOrfuGEA===

**Expected Behavior**:
Roslyn should report `warning CS8602: Dereference of a possibly null reference.` for `c1` and `c3` 

**Actual Behavior**:
Roslyn reports `warning CS8602` for `c2` (edit 11/23/2020: and `c3`)

**Notes**
Since the method `M` ensures that its first argument is not null it's known that conditional access qualifier was not null. Not that Roslyn is aware of this fact as you can check by changing the invocation to `M(c1?.Method(null, null), null)` - Roslyn will not issue a warning for dereferencing `c1` in `c1.ToString()`

Since the conditonal access qualifier is known to be not null it should also be known that the conditional invocation's arguments are executed on all branches that are reachable after the assertion invocation. Roslyn *is not* aware of that as you can check by changing the invocation to `M(c1?.Method(c2 = new C(), null), null);` - Roslyn will issue a warning dereferencing `c2` in `c2.ToString()` even though the expression `c2 = new C()` is known to be executed and it guarantees that `c2` is not null after the call.

Consequently Roslyn fails to pick up the fact that there's an assignment `c1 = null` made _after_ the conditional access check and thus `c1` and `c3` can be null after the assetion call. This is similar to https://github.com/dotnet/roslyn/issues/32328.

As you can see given that the result of `c1?.Method(...)` is guaranteed to be not null it's known that `c1` is not null when it's evaluated as a qualifier to `c1?.Method(...)`. Consequently `c2` cannot be null as it's always assigned with `c1` which is known to be not null in the `c1?.Method(c2 = c1, ...)` expression. `c1` and `c3` actually can be null as `c1` is assigned with `null` in the second argument to `c1?.Method(..., c1 = null)` and is later assigned to `c3`

Roslyn infers nullability completely wrong in this example - issuing a `CS8602` warning for the only variable which cannot be null after the call and fails to warn about possible null dereferences of the other two variables which in fact can contain null values.

**Comment by RikkiGibson** [comment]
Edited the issue description and link to reflect design changes since the issue was created.

I think the warning on 'c2' is by-design because at the point we apply the postcondition from `[NotNull] C? c1`, we can't recover any conditional state that was on the RHS of the `?.` in the argument.

We would basically have to calculate the state once for the purpose of flowing to other arguments to the `M1` call, and then calculate the state again for the purpose of after the call to M1, where we could for example assume that conditional accesses within the argument succeeded, and therefore propagate the not-null state for `c2`.

**Comment by RikkiGibson** [comment]
It feels like we could *introduce* a warning for `c1`, because the maybe-null assignment to it might have occurred. We might want to adjust the `GetSlotsToMarkAsNotNullable` so that we don't climb up past a call, but it's not clear. We likely want to preserve existing behavior on scenarios like `if (obj?.IsGood() == true) { obj.ToString(); }`, without worrying about stuff like `if (obj?.IsGood(obj = null) == true) { obj.ToString(); }`. There might be a solution for this, but fixing this probably won't take priority in the near future.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson you marked thsi as by design.  should this be closed out then?

**Comment by RikkiGibson** [comment]
The above discussion predates improved definite assignment (dotnet/csharplang#4240). When implementing that feature (and applying the corresponding changes to nullable), I decided that this was more in the bug fix category, but that the cost/benefit of fixing it didn't add up.

I think the right thing to do here is close as *won't fix*. If we start seeing new reports of this issue or numerous requests to reconsider then we can do so at that time.
